restarter/hot-restarter.py

```diff
@@ -5,18 +5,69 @@
 import sys
 import time
 
+# The number of seconds to wait for children to gracefully exit after
+# propagating SIGTERM before force killing children.
+# NOTE: If using a shutdown mechanism such as runit's `force-stop` which sends
+# a KILL after a specified timeout period, it's important to ensure that this
+# constant is smaller than the KILL timeout
+TERM_WAIT_SECONDS = 30
+
 restart_epoch = 0
 pid_list = []
 
-def force_kill_all_children():
-  """ Iterate through all known child processes and force kill them. In the future we might consider
-      possibly giving the child processes time to exit but this is fine for now. If someone force kills
-      us and does not clean the process tree this will leave child processes around unless they choose
-      to end themselves if their parent process dies. """
+def term_all_children():
+  """ Iterate through all known child processes, send a TERM signal to each of
+  them, and then wait up to TERM_WAIT_SECONDS for them to exit gracefully,
+  exiting early if all children go away. If one or more children have not
+  exited after TERM_WAIT_SECONDS, they will be forcibly killed """
 
   # First uninstall the SIGCHLD handler so that we don't get called again.
   signal.signal(signal.SIGCHLD, signal.SIG_DFL)
 
+  global pid_list
+  for pid in pid_list:
+    print "sending TERM to PID={}".format(pid)
+    try:
+        os.kill(pid, signal.SIGTERM)
+    except:
+      print "error sending TERM to PID={} continuing".format(pid)
+
+  all_exited = False
+
+  # wait for 30 seconds for children to exit cleanly
+  retries = 0
+  while not all_exited and retries < TERM_WAIT_SECONDS:
+    pid_list_copy = list(pid_list)
+    for pid in pid_list_copy:
+      ret_pid, exit_status = os.waitpid(pid, os.WNOHANG)
+      if ret_pid == 0 and exit_status == 0:
+        # the child is still running
+        continue
+
+      pid_list.remove(pid)
+
+    if len(pid_list) == 0:
+      all_exited = True
+    else:
+      retries += 1
+      time.sleep(1)
```

Comment:
is 1 second too long? If we're here because of an unusual child exit that might be a long time to block our supervisor (e.g. runit) from restarting

Reply:
Seems reasonable to me, I don't see a big win from speeding this up given the amount of time the restart takes anyway.